Feature/validate metric name uniqueness

ycabrer opened the pull request:
<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/master/CONTRIBUTING.md
-->

Metric Names are required to be unique in Scaled Objects

### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO)
- [X] Tests have been added
- [N/A] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [X] Changelog has been updated

Fixes #

ycabrer commented:
@zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).

Also any tips on where to put the documentation update for the metric name validation?

zroubalik commented on controllers/scaledobject_controller.go:
```diff
@@ -214,6 +219,29 @@ func (r *ScaledObjectReconciler) ensureScaledObjectLabel(logger logr.Logger, sca
+	observedMetricNames := make(map[string]struct{})
+	for _, scaler := range scalers {
+		for _, metric := range scaler.GetMetricSpecForScaling() {
+			metricName := metric.External.Metric.Name
```
We should check whether the scaler contains External metrics (`if != nil`), becasue if Resource based Metrics are used, External Metrics are not set.

See: https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/controllers/hpa.go#L135-L162

ycabrer replied:
That's a great point! I went ahead and added a continue when external metrics is nil

zroubalik reviewed (changes requested):
Looking good so far, we should add a check though

zroubalik commented:
> @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).

The static check is strange, you can probably disable this specific one for the finalizers: https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50

> 
> Also any tips on where to put the documentation update for the metric name validation?

We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?

zroubalik commented:
> > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> 
> The static check is strange, you can probably disable this specific one for the finalizers:
> 
> https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> 
> > Also any tips on where to put the documentation update for the metric name validation?
> 
> We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?

@ycabrer any update please? :)

tomkerkhove commented:
> > Also any tips on where to put the documentation update for the metric name validation?
> 
> We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?

Missed this one, sorry @zroubalik, but I agree to have a section on multiple scalers and what to expect of them.

Not sure though if this should include technical details such as metric name validation but more how the decision making will happen maybe?

zroubalik commented:
@ycabrer are you able to update this please? It is completely OK, if you don't have a time to do so. But please let us know, so we can manage this. Thanks!

ycabrer commented:
@zroubalik Sorry for the delay. I've had a busy start to the new year! I'll go ahead and update the docs.

ycabrer commented:
> > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > 
> > 
> > The static check is strange, you can probably disable this specific one for the finalizers:
> > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > 
> > > Also any tips on where to put the documentation update for the metric name validation?
> > 
> > 
> > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> 
> @ycabrer any update please? :)

I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?

zroubalik commented:
> > > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > > 
> > > 
> > > The static check is strange, you can probably disable this specific one for the finalizers:
> > > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > > 
> > > > Also any tips on where to put the documentation update for the metric name validation?
> > > 
> > > 
> > > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> > 
> > 
> > @ycabrer any update please? :)
> 
> I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?

Have you tried adding a new section to the `excluded-rules` section in `.golangci.yaml`?

Something like this:
```yaml
# Excluding interfacer for finalizers, reason: https://github.com/kedacore/keda/pull/1390
 - path: _finalizer.go
      linters:
        - interfacer
```

ycabrer commented:
> > > > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > > > 
> > > > 
> > > > The static check is strange, you can probably disable this specific one for the finalizers:
> > > > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > > > 
> > > > > Also any tips on where to put the documentation update for the metric name validation?
> > > > 
> > > > 
> > > > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> > > 
> > > 
> > > @ycabrer any update please? :)
> > 
> > 
> > I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?
> 
> Have you tried adding a new section to the `exclude-rules` section in `.golangci.yaml`?
> 
> Something like this:
> 
> ```yaml
> # Excluding interfacer for finalizers, reason: https://github.com/kedacore/keda/pull/1390
>  - path: _finalizer.go
>       linters:
>         - interfacer
> ```

That worked 

> > > > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > > > 
> > > > 
> > > > The static check is strange, you can probably disable this specific one for the finalizers:
> > > > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > > > 
> > > > > Also any tips on where to put the documentation update for the metric name validation?
> > > > 
> > > > 
> > > > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> > > 
> > > 
> > > @ycabrer any update please? :)
> > 
> > 
> > I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?
> 
> Have you tried adding a new section to the `exclude-rules` section in `.golangci.yaml`?
> 
> Something like this:
> 
> ```yaml
> # Excluding interfacer for finalizers, reason: https://github.com/kedacore/keda/pull/1390
>  - path: _finalizer.go
>       linters:
>         - interfacer
> ```

That did it! I had chosen the wrong linter.

zroubalik commented on .golangci.yml:
```diff
@@ -56,6 +56,9 @@ issues:
     - path: scale_handler.go
       linters:
         - gocyclo
+    - path: _finalizer.go
```
Could you please add a comment with a description why are we adding this exception, eg. with a link to this PR?

ycabrer replied:
Sure thing! I missed that you had already given an example

zroubalik commented on CHANGELOG.md:
```diff
@@ -40,6 +40,7 @@
 - Improve error reporting in prometheus scaler ([PR #1497](https://github.com/kedacore/keda/pull/1497))
 
 ### Breaking Changes
+- Require metricNames be unique in scaled objects.
```
```suggestion
- Require metricNames be unique in scaled objects ([#1390](https://github.com/kedacore/keda/pull/1390))
```

ycabrer replied:
I just updated it

zroubalik reviewed (approved):
LGTM, thanks @ycabrer!
